### PR TITLE
122 add fig anchor metadata to validations

### LINKED ISSUE #122: Add FIG anchor metadata to validations
We need to provide linkage from our validations to their corresponding section of the FIG. Unfortunately, the HTML anchors are not based on a validation `id` or `name`. For example, the following validation (`E3000`)...

https://github.com/cfpb/regtech-data-validator/blob/485b53b547d161d8dd4758bab2cd1f9da7ff56b7/src/regtech_data_validator/phase_validations.py#L39-L49

...has an anchor of `#4.3.1`.

- SEE: https://www.consumerfinance.gov/data-research/small-business-lending/filing-instructions-guide/2024-guide/#4.3.1

And while we'd prefer to fix the FIG itself, we don't really have time to do that right now. In the short term, we're going to keep track of the anchor values on our side, and return them as part of the validation results.

I think we can name this new field `fig_anchor`, unless anyone has a better name.

So, the work here is:

1. Add the `fig_anchor` tag to each of the validations, with the value from the FIG for each corresponding error/warning.
2. Add `fig_anchor` to the validation results.

There will also be a corresponding issue for https://github.com/cfpb/sbl-filing-api/ so we can return this info through the API.

### PR DESCRIPTION
**Pull request by jcadam14**
Closes #122 

I created a script that read through the fig_source.html, which I saved from browser, and found each of the elements containing the validation ID for each SBLCheck.  Then I grabbed the href and stored that in the fig_anchor in the phase_validations.py file.

I created a pytest to use the same html to loop through both the phase validation fig_anchors and the hrefs to compare to ensure each validation ID had the correct fig anchor.  I also updated the existing test_cli.py formats to include the new fig_anchor field.

Might be a good idea to store this tool somewhere, it's all local currently.  Just in case they change something, it makes it easy to loop through the py file and insert the fig_anchor instead of manually copy/pasting.

Updated the schema to include the fig_anchor in the check, which automatically carries over into the validation results.
Updated the df_to_json to include the fig_anchor in the json we send to clients.  The other df_to's automatically get the fig_anchor.

I don't think we need another story in the filing-api, the fig_anchor for each result will be in the JSON blob.

I would LOVE to figure out how to automatically pull down the FIG html in an actual usable way for the pytests.  However, because it's all javascripted, the actual hrefs don't come across as full links if you do a request.get, or curl, or wget.  The only way I've found to have both the full href link and the Validation ID associated with it is saving the page off in a browser.

### CODE REVIEW

**Comment by hkeeler** [comment]
> I created a script that read through the fig_source.html, which I saved from browser, and found each of the elements containing the validation ID for each SBLCheck. Then I grabbed the href and stored that in the fig_anchor in the file.
> 
> I created a pytest to use the same html to loop through both the phase validation fig_anchors and the hrefs to compare to ensure each validation ID had the correct fig anchor. I also updated the existing test_cli.py formats to include the new fig_anchor field.

🤯 Very cool. 😄 

> Might be a good idea to store this tool somewhere, it's all local currently.

👍 We could add a `utils` to this repo.

> Updated the schema to include the fig_anchor in the check, which automatically carries over into the validation results. Updated the df_to_json to include the fig_anchor in the json we send to clients. The other df_to's automatically get the fig_anchor.

👍 

> I don't think we need another story in the filing-api, the fig_anchor for each result will be in the JSON blob.

👍 

> I would LOVE to figure out how to automatically pull down the FIG html in an actual usable way for the pytests. However, because it's all javascripted, the actual hrefs don't come across as full links if you do a request.get, or curl, or wget. The only way I've found to have both the full href link and the Validation ID associated with it is saving the page off in a browser.

🤔 Maybe something that drives a browser like selenium?

**Comment by hkeeler** [comment]
Here's a quick, dirty, barely-tested scraper. Requires [`beautifulsoup4`](https://pypi.org/project/beautifulsoup4/) and [`requests`](https://pypi.org/project/requests/).

```python
from bs4 import BeautifulSoup, Tag
import requests

url = "https://www.consumerfinance.gov/data-research/small-business-lending/filing-instructions-guide/2024-guide/"
response = requests.get(url)
soup = BeautifulSoup(response.content, 'html.parser')

divs: list[Tag] = soup.find_all("div", class_="o-fig_section__sub-sub")

id_to_anchor = {}

for div in divs:
    a = div.h4.a
    id = a.get('id')
    text = a.text.strip()

    prefix = "Validation ID: "
    
    if text.startswith(prefix):
        validation_id = text.lstrip(prefix)
        id_to_anchor[validation_id] = id

from pprint import pprint
pprint(id_to_anchor)
```

```
{'E0001': '4.1.1',
 'E0002': '4.1.2',
 'E0020': '4.1.3',
 'E0040': '4.1.4',
 'E0060': '4.1.5',
 'E0080': '4.1.6',
 'E0100': '4.1.7',
 'E0120': '4.1.8',
 'E0121': '4.1.9',
 'E0140': '4.1.10',
...
```

**Comment by jcadam14** [comment]
> Here's a quick, dirty, barely-tested scraper. Requires [`beautifulsoup4`](https://pypi.org/project/beautifulsoup4/) and [`requests`](https://pypi.org/project/requests/).
> 
> ```python
> from bs4 import BeautifulSoup, Tag
> import requests
> 
> url = "https://www.consumerfinance.gov/data-research/small-business-lending/filing-instructions-guide/2024-guide/"
> response = requests.get(url)
> soup = BeautifulSoup(response.content, 'html.parser')
> 
> divs: list[Tag] = soup.find_all("div", class_="o-fig_section__sub-sub")
> 
> id_to_anchor = {}
> 
> for div in divs:
>     a = div.h4.a
>     id = a.get('id')
>     text = a.text.strip()
> 
>     prefix = "Validation ID: "
>     
>     if text.startswith(prefix):
>         validation_id = text.lstrip(prefix)
>         id_to_anchor[validation_id] = id
> 
> from pprint import pprint
> pprint(id_to_anchor)
> ```
> 
>  ```
> {'E0001': '4.1.1',
>  'E0002': '4.1.2',
>  'E0020': '4.1.3',
>  'E0040': '4.1.4',
>  'E0060': '4.1.5',
>  'E0080': '4.1.6',
>  'E0100': '4.1.7',
>  'E0120': '4.1.8',
>  'E0121': '4.1.9',
>  'E0140': '4.1.10',
> ...
> ```

The problem with this is it doesn’t provide an actual navigable link, ie a full URL.  I tried this with the request content already and while you can easily get the index (like 4.2.1, etc) I thought the goal was the fig_anchor would be the full https URL path.  If so unfortunately this doesn’t work.  From what I’ve been reading Selenium would be the way to go.  If all we need to do is provide the #4.2.1 and the front end is configured to provide the rest well then yes much easier and a simple request content is doable.

**Comment by jcadam14** [comment]
> > Here's a quick, dirty, barely-tested scraper. Requires [`beautifulsoup4`](https://pypi.org/project/beautifulsoup4/) and [`requests`](https://pypi.org/project/requests/).
> > ```python
> > from bs4 import BeautifulSoup, Tag
> > import requests
> > 
> > url = "https://www.consumerfinance.gov/data-research/small-business-lending/filing-instructions-guide/2024-guide/"
> > response = requests.get(url)
> > soup = BeautifulSoup(response.content, 'html.parser')
> > 
> > divs: list[Tag] = soup.find_all("div", class_="o-fig_section__sub-sub")
> > 
> > id_to_anchor = {}
> > 
> > for div in divs:
> >     a = div.h4.a
> >     id = a.get('id')
> >     text = a.text.strip()
> > 
> >     prefix = "Validation ID: "
> >     
> >     if text.startswith(prefix):
> >         validation_id = text.lstrip(prefix)
> >         id_to_anchor[validation_id] = id
> > 
> > from pprint import pprint
> > pprint(id_to_anchor)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > ```
> > {'E0001': '4.1.1',
> > 'E0002': '4.1.2',
> > 'E0020': '4.1.3',
> > 'E0040': '4.1.4',
> > 'E0060': '4.1.5',
> > 'E0080': '4.1.6',
> > 'E0100': '4.1.7',
> > 'E0120': '4.1.8',
> > 'E0121': '4.1.9',
> > 'E0140': '4.1.10',
> > ...
> > ```
> 
> The problem with this is it doesn’t provide an actual navigable link, ie a full URL. I tried this with the request content already and while you can easily get the index (like 4.2.1, etc) I thought the goal was the fig_anchor would be the full https URL path. If so unfortunately this doesn’t work. From what I’ve been reading Selenium would be the way to go. If all we need to do is provide the #4.2.1 and the front end is configured to provide the rest well then yes much easier and a simple request content is doable.

Actually heck the only piece that changes is that # part so the rest of the url could have just been configured in an env setting and just appending the # stuff at the end.  I was grabbing that in the first 30 minutes of playing around with this lol.  It was trying to get the full actual link that was a pain in the arse.

**Comment by hkeeler** [comment]
Sorry, misunderstood the problem. 

>Actually heck the only piece that changes is that # part so the rest of the url could have just been configured in an env setting and just appending the # stuff at the end. I was grabbing that in the first 30 minutes of playing around with this lol. It was trying to get the full actual link that was a pain in the arse.

Yep. Exactly. 😄

**Comment by jcadam14** [comment]
> Sorry, misunderstood the problem.
> 
> > Actually heck the only piece that changes is that # part so the rest of the url could have just been configured in an env setting and just appending the # stuff at the end. I was grabbing that in the first 30 minutes of playing around with this lol. It was trying to get the full actual link that was a pain in the arse.
> 
> Yep. Exactly. 😄

Man I hate over thinking things lol.  Ok I still have a previous copy of the script that directly used the web page so will be easy to switch back to.  Thanks!

File: tests/data/fig_source.html
Comment (hkeeler) [inline tests/data/fig_source.html:1]: Do we still need this file? Looks like it's not used in the tests anymore.

**Review by hkeeler — COMMENTED** [review]
Looks good. Just the one question about test FIG HTML file.

**Review by hkeeler — APPROVED** [review]

File: tests/data/fig_source.html
Comment (jcadam14) [reply tests/data/fig_source.html:1]: Nope!  Deleted.  Also renamed test_fig_anchors.py to test_fig_links.py just to keep with the 'link' verbiage.  Which I know might come back as fig_url but I'm impatient.

**Pull request merged**